src/database/migrations/0031-create-sources.sql

```diff
@@ -0,0 +1,43 @@
+CREATE TYPE source_type AS ENUM ('data_provider', 'funder', 'organization', 'system');
+
+CREATE TABLE sources (
+  id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
+  source_type source_type NOT NULL,
+  label VARCHAR NOT NULL,
+  organization_id INTEGER REFERENCES organizations(id),
+  funder_id INTEGER REFERENCES funders(id),
+  data_provider_id INTEGER REFERENCES data_providers(id),
+  created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
+  UNIQUE(organization_id, funder_id, data_provider_id),
+  CONSTRAINT chk_proper_fk_for_data_provider_type CHECK (
+    (source_type != 'data_provider') OR
+    (source_type = 'data_provider'
+      AND data_provider_id IS NOT NULL
+      AND funder_id IS NULL
+      AND organization_id IS NULL)
+  ),
+  CONSTRAINT chk_proper_fk_for_funder_type CHECK (
+    (source_type != 'funder') OR
+    (source_type = 'funder'
+      AND data_provider_id IS NULL
+      AND funder_id IS NOT NULL
+      AND organization_id IS NULL)
+  ),
+  CONSTRAINT chk_proper_fk_for_organization_type CHECK (
+    (source_type != 'organization') OR
+    (source_type = 'organization'
+      AND data_provider_id IS NULL
+      AND funder_id IS NULL
+      AND organization_id IS NOT NULL)
+  ),
+  CONSTRAINT chk_proper_fk_for_system_type CHECK (
+    (source_type != 'system') OR
+    (source_type = 'system'
+      AND data_provider_id IS NULL
+      AND funder_id IS NULL
+      AND organization_id IS NULL)
+  )
```

Comment:
This is very complicated, and appears to be a way to maintain consistency between duplicated data: the `source_type` column and the presence of a value in the corresponding FK column (right?). What is the benefit of the `source_type` column?

Can we drop the `source_type` column - and maybe synthesize it in query responses if it's very helpful to have on the application side - and use something like the construction [suggested here](https://github.com/PhilanthropyDataCommons/service/pull/1116#discussion_r1700672988)?

Synthesizing it might look like
```SQL
SELECT ...,
  CASE WHEN data_provider_id IS NOT NULL THEN 'data_provider'
    WHEN funder_id IS NOT NULL THEN 'funder'
    WHEN organization_id IS NOT NULL THEN 'organization'
    ELSE 'system'
  END AS source_type,
...
```
and the constraint might look like
```SQL
  CONSTRAINT chk_only_one_fk CHECK (
    num_nonnulls(organization_id, funder_id, data_provider_id) <= 1
  )

Reply:
Maybe this would do the trick: a `generated_column` could I think give us everything -- we could have the simpler constraints, while also making the type explicit in the data / avoiding the issue of putting the burden of translating shapes to types on the query writers.

Reply:
I implemented with a generated column -- it is a little awkward to have our insert logic take a source_type without actually writing it.

What I mean is that now the logic mapping source type to <=> from fk relationship shape is now duplicated in both the `insertOne.sql` as well as the table definition, as opposed to it being that the mapping is in the insertion, and the constraints for validating data consistency are in the table.

That said, not a huge deal.

Reply:
Talked with @jasonaowen about this all!  Here's where things landed:

let's take source_type out entirely (in the type as well as the table)

Everything is simpler, and clients can handle checking which value is populated themselves.

We're also going to ditch the idea of a "system" type and instead have PDC become a data provider.